### PR TITLE
Video and audio recorder implementation

### PR DESCRIPTION
**Pull request by nirvn**
This PR implements video as well as recording functionality straight within the QField feature form through the external resource/attachment editor widget.

image

_Notice the brand new video camera and microphone icons_

Implementation notes on video recording:
- on Android, native video recording is possible (*)
- on other platforms, QField's QML-based camera widget has received some love and is now able to record video as well

_* code for native video recording on iOS is in there too, but I have not tested it and I have very little experience/knowledge on Objective-C_

Implementation notes on audio recording:
- the audio recording is done via a brand new QML audio recorder widget (see GIF below); you simply click on the big record-red button to record, then click on the OK/check button to save it
- when recording, users are provided with level feedback via a growing/shrinking circle under the record button (*)

_* level feedback available on platforms where QAudioProbe is supported; when probing isn't possible, a simple animation loops instead_

Peek 2023-02-19 17-53

### CODE REVIEW

**Comment by qfield-fairy** [comment]
🎉 Ta-daaa, freshly created APKs are available for 5df5c75ca1372698fa58397c233d395381f6ff13: [**arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-arm64-android.apk)
<details>
<summary>Other architectures</summary>

- [arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-arm-android.apk)
- [x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-x64-android.apk)
- [x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-x86-android.apk)
- [**all access arm64-android**](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-arm64-android.apk)
- [all access arm-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-arm-android.apk)
- [all access x64-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-x64-android.apk)
- [all access x86-android](https://sos-ch-dk-2.exo.io/qfieldapks/ci-builds/qfield_all_access-dev-3924-cfb9415aaea0fb962f0595613fca2b9ce72a1c3a-x86-android.apk)
</details>

File: src/core/audiorecorder.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+#ifndef AUDIORECORDER_H
+#define AUDIORECORDER_H
+
+#include <QObject>
+#if QT_VERSION < QT_VERSION_CHECK( 6, 0, 0 )
+#include <QAudioProbe>
+#include <QAudioRecorder>
+
+class AudioRecorder : public QAudioRecorder
```
Comment (nirvn) [inline src/core/audiorecorder.h:25]: Note to self/TODO: documentation

File: src/core/platforms/platformutilities.cpp [hunk 2]
```diff
@@ -239,6 +239,15 @@ ResourceSource *PlatformUtilities::getCameraPicture( QQuickItem *parent, const Q
   return nullptr;
 }
 
+ResourceSource *PlatformUtilities::getCameraVideo( QQuickItem *parent, const QString &prefix, const QString &videoFilePath, const QString &suffix )
```
Comment (m-kuhn) [inline src/core/platforms/platformutilities.cpp:242]: Todo: capabilities?

Comment (nirvn) [reply src/core/platforms/platformutilities.cpp:242]: @m-kuhn , yeah, planned in a separate PR.

File: src/core/platforms/platformutilities.cpp [hunk 3]
```diff
@@ -291,7 +304,7 @@ ResourceSource *PlatformUtilities::getFile( QQuickItem *parent, const QString &p
   switch ( fileType )
   {
     case AudioFiles:
-      filter = tr( "Audio files (*.mp3 *.aac *.ogg)" );
+      filter = tr( "Audio files (*.mp3 *.aac *.ogg *.m4a *.mp4 *.mov)" );
```
Comment (m-kuhn) [inline src/core/platforms/platformutilities.cpp:307]: Isn't .mov a video file?

**Review by m-kuhn — APPROVED** [review]

File: src/core/platforms/platformutilities.cpp (hunk 3 above)
Comment (nirvn) [reply src/core/platforms/platformutilities.cpp:307]: .mov is a container which can hold video and/or audio. For e.g., Qt was doing audio-only .mov using the QAudioRecorder before setting it to ogg.

**Pull request merged**